cranelift/codegen/src/isa/x64/inst/emit.rs

```diff
@@ -437,7 +437,33 @@ pub(crate) fn emit(
                 OperandSize::Size64 => (0xF7, LegacyPrefixes::None),
             };
 
-            sink.add_trap(TrapCode::IntegerDivisionByZero);
+            // If div traps are avoided then no trap is registered here since
+            // the instruction should never trap due to lowering rules.
+            //
+            // Otherwise though this instruction is allowed to trap. There's
+            // one trapping condition for unsigned division, which is
+            // always divide-by-zero, so that's reported here. There are two
+            // conditions for signed division, though, either divide-by-zero
+            // or integer overflow with INT_MIN/-1. Lowering explicitly
+            // checks for 0 because it's easier and defers the overflow
+            // detection to happening via a trap.
+            //
+            // Note, though, that this instruction is also used as part of the
+            // `CheckedSRemSeq*` lowering which uses signed division but
+            // statically rules out the integer overflow case, in which case
+            // the listed trap here is divide-by-zero. This means that only the
+            // actual CLIF `sdiv` instruction ends up triggering the integer
+            // overflow case.
```

Comment:
I wonder if it would be simpler to include a `trap_code` field in the instruction, and set this when creating it in the lowering? That would avoid the implicit state flow via the flag on the `EmitState` and would also place this conditional logic closer to where the rest of the instruction sequence is specified.

Reply:
Ah good suggestion!